Source/Particles/Collision/BinaryCollision/DSMC/SplitAndScatterFunc.H

```diff
@@ -9,173 +9,287 @@
-    ptile2.resize(ptile2_index + num_added);
+    /**
+     * \brief Function that performs the particle scattering and injection due
+     * to binary collisions.
+     *
+     * \return num_added the number of particles added to each species.
+     */
+    AMREX_INLINE
+    amrex::Vector<int> operator() (
+        const index_type& n_total_pairs,
+        // Tile& ptile1, Tile& ptile2,
+        const SoaData_type& /*soa_1*/, const SoaData_type& /*soa_2*/,
+        const amrex::Vector<WarpXParticleContainer*>& pc_products,
+        ParticleTileType** AMREX_RESTRICT tile_products,
+        const amrex::ParticleReal m1, const amrex::ParticleReal m2,
+        const amrex::Vector<amrex::ParticleReal>& /*products_mass*/,
+        const index_type* AMREX_RESTRICT mask,
+        const amrex::Vector<index_type>& products_np,
+        const SmartCopy* AMREX_RESTRICT copy_species1,
+        const SmartCopy* AMREX_RESTRICT copy_species2,
+        const index_type* AMREX_RESTRICT p_pair_indices_1,
+        const index_type* AMREX_RESTRICT p_pair_indices_2,
+        const amrex::ParticleReal* AMREX_RESTRICT p_pair_reaction_weight ) const
+    {
+        using namespace amrex::literals;
 
-    const auto ptile1_data = ptile1.getParticleTileData();
-    const auto ptile2_data = ptile2.getParticleTileData();
+        if (n_total_pairs == 0) { return amrex::Vector<int>(m_num_product_species, 0); }
 
-    ParallelForRNG(n_total_pairs,
-    [=] AMREX_GPU_DEVICE (int i, RandomEngine const& engine) noexcept
-    {
-        if (mask[i])
+        amrex::Gpu::DeviceVector<index_type> offsets(n_total_pairs);
+        index_type* AMREX_RESTRICT offsets_data = offsets.data();
+        const index_type* AMREX_RESTRICT p_offsets = offsets.dataPtr();
+
+        // The following is used to calculate the appropriate offsets. Note that
+        // a standard cummulative sum is not appropriate since the mask is also
+        // used to specify the type of collision and can therefore have values >1
+        auto const total = amrex::Scan::PrefixSum<index_type>(n_total_pairs,
+            [=] AMREX_GPU_DEVICE (index_type i) -> index_type { return mask[i] ? 1 : 0; },
+            [=] AMREX_GPU_DEVICE (index_type i, index_type s) { offsets_data[i] = s; },
+            amrex::Scan::Type::exclusive, amrex::Scan::retSum
+        );
+
+        amrex::Vector<int> num_added_vec(m_num_product_species);
+        for (int i = 0; i < m_num_product_species; i++)
+        {
+            // How many particles of product species i are created.
+            const index_type num_added = total * m_num_products_host[i];
+            num_added_vec[i] = static_cast<int>(num_added);
+            tile_products[i]->resize(products_np[i] + num_added);
+        }
+
+        // this works for DSMC since the colliding particles are also products
+        const auto soa_1 = tile_products[0]->getParticleTileData();
+        const auto soa_2 = tile_products[1]->getParticleTileData();
```

Comment:
I wanted to highlight this bit of code. For some reason that I don't understand it doesn't work to set the id of colliding particles to invalid using the `soa_1` object passed into this function (particles are not actually removed from the simulation in that case). However, using the tiles to get the SoA values as is done here does produce the expected result. Based on the similarity between this code and the fusion module, I expect the same issue to exist there. I will submit a follow-up PR once this is merged to change the fusion approach as well.